contrib/pg_tde/src/access/pg_tde_tdemap.c

```diff
@@ -925,6 +935,84 @@ pg_tde_move_rel_key(const RelFileLocator *newrlocator, const RelFileLocator *old
 	pfree(enc_key);
 }
 
+/* It's called by seg_write in crit section so no pallocs, hence
+ * need keyfile_path
+ */
+void
+pg_tde_wal_last_key_set_lsn(XLogRecPtr lsn, const char *keyfile_path)
+{
+	LWLock	   *lock_pk = tde_lwlock_enc_keys();
+	int			fd = -1;
+	off_t		write_pos,
+				last_key_idx,
+				prev_key_pos;
+	InternalKey	prev_key;
+
+	fd = BasicOpenFile(keyfile_path, O_RDWR | PG_BINARY);
+	if (fd < 0)
+	{
+		ereport(ERROR,
+				(errcode_for_file_access(),
+				 errmsg("could not open tde file \"%s\": %m",
+					keyfile_path)));
+	}
+
+	last_key_idx = ((lseek(fd, 0, SEEK_END) - TDE_FILE_HEADER_SIZE) / INTERNAL_KEY_DAT_LEN) - 1;
+	write_pos = TDE_FILE_HEADER_SIZE + (last_key_idx * INTERNAL_KEY_DAT_LEN) + offsetof(InternalKey, start_lsn);
+
+	LWLockAcquire(lock_pk, LW_EXCLUSIVE);
+	/* TODO: pgstat_report_wait_start / pgstat_report_wait_end */
+	if (pg_pwrite(fd, &lsn, sizeof(XLogRecPtr), write_pos) != sizeof(XLogRecPtr))
```

Comment:
We have an inconsistency here: the internal keys for xlog are normally are xlogged, but not the modifications in this function. Can't this cause issues?

Reply:
My reasoning for why no XLog needed here:
1. We should not XLog it to replicas as they might diverge in terms of internal keys. The user can turn on and off wal encryption on replicas notwithstanding the primary.
2. As for reliability, this write/update itself acts as the XLog. We record new start_lsn on disk and only then "use it". So if we crash before the start_lsn update - no side effects and the next write will try to do the same. If we crash after the update but before the WAL write - it's ok, as start_lsn indicates that everything >= is affected by the key, so the WAL writes after restart may just continue. If it crashes after the start_lsn and the user changes settings - a new key will be generated and the old one won't matter anymore.